core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ElementDestinationSelector.java

```diff
@@ -0,0 +1,143 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.impl.domain.variable.descriptor.ListVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableDemand;
+import org.optaplanner.core.impl.domain.variable.index.IndexVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
+import org.optaplanner.core.impl.domain.variable.inverserelation.SingletonListInverseVariableDemand;
+import org.optaplanner.core.impl.domain.variable.supply.SupplyManager;
+import org.optaplanner.core.impl.heuristic.selector.AbstractSelector;
+import org.optaplanner.core.impl.heuristic.selector.IterableSelector;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.value.EntityIndependentValueSelector;
+import org.optaplanner.core.impl.solver.random.RandomUtils;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+
+/**
+ * Selects destinations for list variable change moves. The destination specifies a future position in a list variable,
+ * expressed as an {@link ElementRef}, where a moved element or subList can be inserted.
+ * <p>
+ * Destination completeness is achieved by using both entity and value child selectors.
+ * When an entity <em>A</em> is selected, the destination becomes <em>A[0]</em>.
+ * When a value <em>x</em> is selected, its current position <em>A[i]</em> is determined using inverse and index supplies and
+ * the destination becomes <em>A[i + 1]</em>.
+ * <p>
+ * Fairness in random selection is achieved by first deciding between entity and value selector with a probability that is
+ * proportional to the entity/value ratio. The child entity and value selectors are assumed to be fair.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class ElementDestinationSelector<Solution_> extends AbstractSelector<Solution_>
+        implements IterableSelector<Solution_, ElementRef> {
+
+    private final ListVariableDescriptor<Solution_> listVariableDescriptor;
+    private final EntitySelector<Solution_> entitySelector;
+    private final EntityIndependentValueSelector<Solution_> valueSelector;
+    private final boolean randomSelection;
```

Comment:
Looks to me as if this should implement equality on these four fields.
Selector equality is newly important, as we now have the supply-demand nearby/pillar cache which depends on it.

Reply:
I don't see it as necessary at this point but I'll keep it in mind. IIUC, each component of a demand must implement equality, so if it turns out that `ElementDestinationSelector` goes into the nearby distance matrix demand, I'll added. Otherwise YAGNI. For example, DefaultSubChainSelector does not implement it either.

Reply:
Right. Just keep in mind that some of these selectors go into those demands transitively, as properties of other selectors.